[Goel Akshit] iP

goel-a opened the pull request:
##### Duke[](https://github.com/goel-a/ip)

> “Dedication makes dreams come true.” – Kobe Bryant ([source](https://parade.com/1234641/jessicasager/kobe-bryant-quotes/))

Duke frees your mind of having to remember things you need to do and needing to do them. It's,

-   text-based
-   easy to learn
-   ~~FAST~~  *SUPER FAST* to use

All you need to do is,

1.  download it from  [here](https://github.com/goel-a/ip/releases/tag/v0.1).
2.  double-click it.
3.  add your tasks.
4.  let it manage your tasks for you 😉

And it is  **FREE**!

Features:

 - [x] Managing tasks
 - [x] Managing deadlines

----------

If you're a Java programmer, you can use it to practice Java too. Here's the  `main`  method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(Main.class, args);
    }
}
```

goel-a committed:
Level-1 Implemented

goel-a committed:
Added the ability to store whatever text entered by the user and display them back to the user when requested.

goel-a committed:
Add the ability to mark tasks as done and the ability to change the status back to not done.

goel-a committed:
Add support for tracking three types of tasks

goel-a committed:
Test using the I/O redirection technique

goel-a committed:
Add error handling

goel-a committed:
Add support for deleting tasks from the list.

goel-a committed:
Fixed Exception Handling by adding more Specific Exceptions

goel-a committed:
Add ability to save tasks to hard disk

goel-a committed:
Add ability to store Tasks with LocalDate Objects

goel-a committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Add ability to store Tasks with LocalDate Objects

# Conflicts:
#	src/main/java/TaskManager.java

goel-a committed:
Fixed merge errors

goel-a committed:
Made the project more OOP

goel-a committed:
Organize the classes into suitable java packages

goel-a committed:
Add JUnit tests to test the behavior of the code

goel-a committed:
Package the app as an executable JAR file so that it can be distributed easily

goel-a committed:
Data File bug Fixed

goel-a committed:
Added data file

goel-a committed:
Partial javadocs

goel-a committed:
Add Javadocs to all classes

goel-a committed:
Edit code to comply with coding standard

goel-a committed:
Add find feature

goel-a committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Edit code to comply with coding standard

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/utils/DukeException.java
#	src/main/java/duke/utils/TaskList.java
#	src/main/java/duke/utils/Ui.java

goel-a committed:
Minor fix

goel-a committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Add find feature

# Conflicts:
#	src/main/java/duke/utils/DukeException.java
#	src/main/java/duke/utils/Ui.java

goel-a committed:
Add gradle support for builds

goel-a committed:
Add checkstyle configs

goel-a committed:
Partial GUI implementation

goel-a committed:
Add GUI

tobihy commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,40 @@
+     * @param isComp Completion Status of Deadline
+     * @param date Date by which Deadline has to be Completed
+     */
+    public Deadline(String desc, boolean isComp, LocalDate date) {
```
Perhaps a more descriptive parameter name, such as `isCompleted` would be better? I noticed the same issue in other places too.

tobihy commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,87 @@
+ */
+public abstract class Task implements Serializable {
+
+    /** Store a string description of the task **/
```
Just a minor nit, but JavaDoc comments need not be in imperative mood. I noticed similar comments in other parts too.

tobihy commented on src/main/java/duke/utils/Duke.java:
```diff
@@ -0,0 +1,43 @@
+package duke.utils;
+import duke.utils.*;
```
As per the coding standards, maybe adjust the setting in your IDE such that wildcard imports are not automatically applied?

tobihy commented on src/main/java/duke/utils/DukeException.java:
```diff
@@ -0,0 +1,93 @@
+        }
+    }
+
+
+    /**
+     * Handle the exception where the user enters an
+     * invalid index for a action on the list
+     */
+    public static class DukeInvalidNumberException extends DukeException {
+
+        /**
+         * Return error message telling user that
+         * the index entered is invalid
+         *
+         * @return String representation of error message
+         */
+        @Override
+        public String toString() {
+            return super.toString() + "PLEASE ENTER A VALID NUMBER TO MARK/UNMARK/DELETE";
+        }
+    }
+
+
+    /**
+     * Handle the exception where the user doesn't enter a
+     * date when adding a new event/deadline
+     */
+    public static class DukeNoTimeProvided extends DukeException {
+
+        /**
+         * Return error message telling user that
+         * no date has been provided
+         *
+         * @return String representation of error message
+         */
+        @Override
+        public String toString() {
+            return super.toString() + "PLEASE ENTER A VALID DATE FOR THE EVENT/DEADLINE";
+        }
+    }
```
Maybe consider extracting out such exceptions into their separate classes to enhance readability?

tobihy commented on src/test/java/ToDoTest.java:
```diff
@@ -0,0 +1,16 @@
+    public void testCompletionStatus(){
+        ToDo task = new ToDo("Work", false);
+        task.markCompleted();
+        assertTrue(task.getCompletionStatus());
+        task.markNotCompleted();
+        assertFalse(task.getCompletionStatus());
```
Perhaps split the two JUnit tests into two separate test methods?

tobihy commented on src/main/java/duke/utils/Parser.java:
```diff
@@ -0,0 +1,123 @@
+     */
+    public static String parse(String userInput, TaskList tl) throws DukeException {
+
+        StringTokenizer st = new StringTokenizer(userInput, " ");
+        String curr = st.nextToken();
```
Interesting way to parse user input 👍

tobihy reviewed (commented):
Nice display photo for Duke!

Overall, I found your code easy to read and understand. I liked the unique way you parsed user inputs too. However, I found some minor problems in terms of nested classes and variable/parameter names that could be improved on further. In addition, certain JUnit test methods could be refactored for better readability. Other than that, keep up the good work! 👍

xpzmichael commented on src/main/java/duke/utils/Parser.java:
```diff
@@ -0,0 +1,123 @@
+
+            return tl.printList();
+
+            case "mark":
```
indentation mistake.

xpzmichael commented on src/main/java/duke/utils/Parser.java:
```diff
@@ -0,0 +1,123 @@
+
+        case "bye":
+
+            return Ui.printBye();
```
if user entered **_"bye do something /at Sunday"_**, your program will still end. Maybe it would be better to recognize this situation as **an invalid user input** rather than quitting directly.

xpzmichael commented on src/main/java/duke/utils/Storage.java:
```diff
@@ -0,0 +1,56 @@
+
+import duke.task.Task;
+
+import java.io.*;
```
Wildcard import statement.

xpzmichael reviewed (commented):
Nice code! Very well written and well organized. Maybe parser class can be more presice in the sense that "bye do something" will not terminate the program directly, but recognize it as input typo or do a double confirm instead.

goel-a committed:
Add guard clause in Storage.java:

 There is an issue where, upon executing the jar file, the user is unable to
store his list of tasks when the program closes, due to filepath issues

Fixed this issue by checking for file directory when closing the program.

goel-a committed:
Use of Assertions

 Used assertions to catch potential bugs in the program. Majorly used in
the Parser and Duke classes to ensure that user input is not null.

 Added assertions with error messages.

goel-a committed:
Minor variable name fixes

goel-a committed:
Code Quality Improvements

 Following the various practices explained in the module textbook, improvements
to multiple sections of the code was made. Particularly in the Parser class, where
methods were extracted to follow SLAP principles.

goel-a committed:
Merge pull request #1 from goel-a/branch-A-Assertions

Use of Assertions

goel-a committed:
Merge branch 'master' into branch-A-CodeQuality

goel-a committed:
Merge pull request #2 from goel-a/branch-A-CodeQuality

Code Quality Improvements

goel-a committed:
C-Sort

 Added the ability to allow the user to sort events/deadlines by date.

 Did this by making the event and deadline classes comparable, so that
they could be sorted in an Arraylist and displayed to the user.

goel-a committed:
Javadoc, checkstyle and minor bug fixes

goel-a committed:
Add product screenshot

goel-a committed:
Updated README with User Guide

goel-a committed:
Update README.md

goel-a committed:
Update README.md

goel-a committed:
Update README.md

goel-a committed:
Set theme jekyll-theme-hacker

goel-a committed:
Update README.md

goel-a committed:
Code quality fixes based on issue tracker

goel-a committed:
Merge branch 'master' of https://github.com/goel-a/ip

* 'master' of https://github.com/goel-a/ip:
  Update README.md
  Set theme jekyll-theme-hacker
  Update README.md
  Update README.md
  Update README.md
